src/OrleansAzureUtils/MultiClusterNetwork/AzureTableBasedGossipChannel.cs

```diff
@@ -0,0 +1,187 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime.Configuration;
+using Orleans.MultiCluster;
+
+namespace Orleans.Runtime.MultiClusterNetwork
+{
+    // An implementation of a gossip channel based on a  standard orleans azure table
+    // multiple gossipnetworks can use the same table, and are separated by pkey = GlobalServiceId
+    internal class AzureTableBasedGossipChannel : IGossipChannel
+    {
+
+        private TraceLogger logger;
+        private GossipTableInstanceManager tableManager;
+        private static int seqno;
+
+        public string Name { get; private set; }
+
+        public async Task Initialize(GlobalConfiguration globalconfig, string connectionstring)
+        {
+            Name = "AzureTableBasedGossipChannel-" + ++seqno;
+            logger = TraceLogger.GetLogger(Name, TraceLogger.LoggerType.Runtime);
+
+            logger.Info("Initializing Gossip Channel for ServiceId={0} using connection: {1}, SeverityLevel={2}",
+                globalconfig.GlobalServiceId, ConfigUtilities.RedactConnectionStringInfo(connectionstring), logger.SeverityLevel);
+
+            tableManager =
+                await GossipTableInstanceManager.GetManager(globalconfig.GlobalServiceId, connectionstring, logger);
+        }
+
+        // used by unit tests
+        public Task DeleteAllEntries()
+        {
+            logger.Info("DeleteAllEntries");
+            return tableManager.DeleteTableEntries();
+        }
+
+        public struct Pair<L,R>
```

Comment:
but you don't really need mutation, correct? I'm seeing that you only mutate in UpdateDictionaryRight, but effectively getting the value from the dictionary, "mutating" that local copy of the value (that no one else sees mutated since it's a value object), and then re-assigning to the dictionary... instead of mutating that local value you can just create a new instance of KeyValuePair with that value already mutated and storing that in the dictionary.
Having said that, I was just trying to understand. Maybe naming-wise `KeyValuePair` sounds odd in your scenario where it does not act as a key and a value... But having immutable semantics is more maintainable though, as it's a lot easier to comprehend than what's here where you can harmlessly mutate variables in some scenarios because of of the technicality that it's a struct (but of course the definition of whether it's a struct or class is not close to where it is used). If it was immutable, then whether this is a struct or a class is a choice made for performance, but the usage itself would be equivalent.